lib/algolia.ex

```diff
@@ -196,119 +196,129 @@ defmodule Algolia do
+  def add_objects(index, objects, id_attribute_opts \\ [], request_options \\ [])
+
+  def add_objects(index, objects, [id_attribute: id_attribute], request_options) do
+    save_objects(index, objects, [id_attribute: id_attribute], request_options)
   end
 
-  @doc """
-  Add multiple objects
-  """
-  def add_objects(index, objects) do
+  def add_objects(index, objects, request_options, []) do
     objects
     |> build_batch_request("addObject")
-    |> send_batch_request(index)
-  end
-
-  @doc """
-  Add multiple objects, with an attribute as objectID
-  """
-  def add_objects(index, objects, id_attribute: id_attribute) do
-    save_objects(index, objects, id_attribute: id_attribute)
+    |> send_batch_request(index, request_options)
   end
 
   @doc """
   Save a single object, without objectID specified, must have objectID as
   a field
   """
-  def save_object(index, object, id_attribute: id_attribute) do
-    object_id = object[id_attribute] || object[to_string(id_attribute)]
+  def save_object(index, object, opts \\ [id_attribute: :objectID], request_options \\ [])
 
-    if !object_id do
-      raise ArgumentError,
-        message: "Your object #{object} does not have a attribute #{id_attribute}"
-    end
+  def save_object(index, object, opts_or_id, request_options) when is_map(object) do
+    object_id = object_id_for_save!(object, opts_or_id)
 
-    save_object(index, object, object_id)
-  end
-
-  def save_object(index, object, object_id) when is_map(object) do
     body = Jason.encode!(object)
     path = "#{index}/#{object_id}"
 
     :write
-    |> send_request(%{method: :put, path: path, body: body})
+    |> send_request(%{method: :put, path: path, body: body, options: request_options})
     |> inject_index_into_response(index)
   end
 
-  def save_object(index, object) when is_map(object) do
-    object_id = object["objectID"] || object[:objectID]
-
-    if !object_id do
+  defp object_id_for_save!(object, id_attribute: :objectID) do
+    type_insensitive_id(object, :objectID) ||
       raise ArgumentError,
         message: "Your object must have an objectID to be saved using save_object"
-    end
+  end
 
-    body = Jason.encode!(object)
-    path = "#{index}/#{object_id}"
+  defp object_id_for_save!(object, id_attribute: id_attribute) do
+    type_insensitive_id(object, id_attribute) ||
+      raise ArgumentError,
+        message: "Your object #{object} does not have a '#{id_attribute}' attribute"
+  end
 
-    :write
-    |> send_request(%{method: :put, path: path, body: body})
-    |> inject_index_into_response(index)
+  defp object_id_for_save!(_object, id), do: id
+
+  defp type_insensitive_id(object, id_attribute) do
+    object[id_attribute] || object[to_string(id_attribute)]
   end
 
   @doc """
   Save multiple objects
   """
-  def save_objects(index, objects, id_attribute: id_attribute) when is_list(objects) do
+  def save_objects(index, objects, id_attribute_opts \\ [], request_options \\ [])
```

Comment:
Still thinking if it's better to increase arity to 4 here or reuse the original keyword list.

Reply:
Decided to keep the same arity on functions that already accept a keyword list.